Display date added and modified for audio contributions on sentence page

cblanken opened the pull request:
This PR addresses issue #2275. 
To keep it simple, the created date and last modified date are both displayed. However, I think it might be better to only display the modified date if it differs from the creation date. It might also be better to calculate the days/months/years "ago" that the audio recording was created or modified to match the Logs section below it.

Let me know what you think. I'm happy to make any necessary changes.

## Audio contributions
![image](https://github.com/Tatoeba/tatoeba2/assets/19908880/32ae2547-909f-4bf2-a56a-78552c193ee5)

## Logs section for reference
![image](https://github.com/Tatoeba/tatoeba2/assets/19908880/52137fc5-888f-4f51-b537-f42783358fe2)

Yorwba commented:
Thank you for the PR!

For the dates I would use the [`ago` function from the `DateHelper`](https://github.com/Tatoeba/tatoeba2/blob/746693160c44c100560532631ce8d6478edd7ff1/src/View/Helper/DateHelper.php#L109), which is used in the logs section. ([`getDateLabel`](https://github.com/Tatoeba/tatoeba2/blob/746693160c44c100560532631ce8d6478edd7ff1/src/View/Helper/DateHelper.php#L79) is also an option but doesn't show a full sentence if creation and modification dates are the same.)

The CSS needs some protection against overflow if the localized string gets too long:
![Date adddd...dded](https://github.com/Tatoeba/tatoeba2/assets/6555947/b2d6b43a-1bb3-4d21-9476-0d66d2230d1c)

cblanken commented:
I've updated the CSS to handle overflows and setup the date display to use the DateHelper's `ago` function.

One note though. Since the audio section rendering is handled in Angular. This makes it a bit awkward to use the DateHelper, so I added a couple fields to the audios array before it's encoded into JSON and handed off to the client. However, the existing version of `ago()` embeds the `&nbsp;` HTML entity as part of the output which, as far as I know, can only be used if we adapt the audios section to be rendered server-side with PHP.

Instead, I modified the `ago()` function to embed the non-breaking spaces as a Unicode codepoint so it could render properly via Angular.

I don't think that should be a problem for rendering elsewhere, but I just wanted to point it out as it does cause a few tests to fail which seem to expect the `&nbsp;`.

jiru commented on src/Model/Table/SentencesTable.php:
```diff
@@ -665,7 +665,7 @@ public function fields($what = [])
     public function contain($what = [])
     {
         $audioContainment = function (Query $q) use ($what) {
-            $q = $q->select(['id', 'external', 'sentence_id']);
```
I think this should be moved to the `if (isset($what['sentenceDetails'])) {` block below in a way that adds `created` and `modified` fields to the `select` operation. This way we only retrieve these extra fields on sentence pages (as opposed to retrieving them everywhere a sentence block is displayed).

cblanken replied:
See commit ba8e93d

jiru commented on src/Template/Element/sentences/audio.ctp:
```diff
@@ -96,6 +98,14 @@ $this->AngularTemplate->addTemplate(
+            <div class="timestamp">
+                <div><?= format(__('Last modified')) ?></div>
+                <div class="since"><?= format(__('{date}'), [ 'date' => '{{ audio.modified_ago }}' ]) ?></div>
+            </div>
```
I’d like to suggest not displaying the "Last modified" section at all when the "last modified" date is the same as the "Added" one. That would reduce redundancy as I guess these dates are the same for most audio recordings.

jiru commented:
The display of the dates look really nice! I could be made even nicer if you check how it looks when you login as admin. There are some extra information below the dates that could benefit from having the same look (with sections separated by a horizontal bar).

cblanken commented on src/Template/Element/sentences/audio.ctp:
```diff
@@ -96,6 +98,14 @@ $this->AngularTemplate->addTemplate(
+                <div class="since"><?= format(__('{date}'), [ 'date' => '{{ audio.created_ago }}' ]) ?></div>
+            </div>
+            <div class="timestamp">
+                <div><?= format(__('Last modified')) ?></div>
```
See commit dec4e68

cblanken replied:
See commit dec4e68

cblanken commented on src/Template/Element/sentences/audio.ctp:
```diff
@@ -96,6 +98,14 @@ $this->AngularTemplate->addTemplate(
+            </div>
+            <div class="timestamp">
+                <div><?= format(__('Last modified')) ?></div>
+                <div class="since"><?= format(__('{date}'), [ 'date' => '{{ audio.modified_ago }}' ]) ?></div>
```
See commit dec4e68

cblanken commented:
Screenshots as of ba8e93d

## Contributor
![image](https://github.com/Tatoeba/tatoeba2/assets/19908880/abad7c36-a196-4dbb-ac29-fe08254b68bf)

## Admin
![image](https://github.com/Tatoeba/tatoeba2/assets/19908880/c3458024-cea0-4c5b-9c35-6fa14f1cac47)

jiru commented on src/Model/Table/SentencesTable.php:
```diff
@@ -665,14 +665,17 @@ public function fields($what = [])
-            return $q->contain(['Users' => ['fields' => $usersFields]]);
+
+            $q = $q->select($audioFields);
+            return $q->contain(array_merge(['Users' => ['fields' => $usersFields]]));
```
I believe the call to `array_merge()` can be removed. :wink:

jiru commented on src/View/Helper/DateHelper.php:
```diff
@@ -137,13 +137,13 @@ public function ago($date, $alone=true)
         } elseif ($diff->days > 0) {
-            return format(__n('yesterday', '{n}&nbsp;days ago', $diff->days), array('n' => $diff->days));
+            return format(__n('yesterday', "{n}\u{00A0}days ago", $diff->days), array('n' => $diff->days));
         } elseif ($diff->h > 0) {
-            return format(__n('an hour ago', '{n}&nbsp;hours ago', $diff->h), array('n' => $diff->h));
+            return format(__n('an hour ago', "{n}\u{00A0}hours ago", $diff->h), array('n' => $diff->h));
         } else {
             // we stop at minute accuracy
             $minutes = max($diff->i, 1);
-            return format(__n('a minute ago', '{n}&nbsp;minutes ago', $minutes), array('n' => $minutes));
```
Thanks, I see! The existing code base mostly assumes translations can contain HTML, as it is sometimes necessary to include formatting tags such as `<a>`, `<em>` or `<strong>`.

So I am concerned that this change would invalidate a lot of translations that won’t be added back before long in many languages since we don’t have a dedicated team of UI translators. For example, let’s say a contributor translated the UI up to 100% in Korean (just an example right) and then does not actively maintain the translation, so every time some new translatable strings are added or replaced, the translation coverage decreases to 99%, 98%, 95% etc. These date strings are quite visible because they are used in a lot of places, so updating them will have a substantial impact on Tatoeba’s outlook in many UI languages for some time.

I am also concerned about the discrepancy it creates. For example some translators may nevertheless add some `&nbsp;` in their translation of "{n} days ago" and see that it works it one place, such as the Wall, and conclude it’s fine, whereas it’s not displaying correctly in the audio block.

From a security standpoint it’s not very secure to allow HTML from translations, so what AngularJS is doing by default is actually the right thing to do. But this is not the way we deal with translations for now on Tatoeba, so I suggest using AngularJS’s `trustAsHtml()` instead to keep things consistent between AngularJS code and PHP code.

cblanken replied:
Ah, okay. I wasn't aware it would break translations like that. I'm not sure how I feel about blindly trusting the translations with `trustAsHtml()`. I feel like it's overkill since the only issue is a single html entity. Since this is an isolated incident, what if the `&nbsp;` entity is instead replaced with the Unicode NBSP just before Angular renders them. Like this.

### audio-details.ctrl.js
```js
function init(audios, audioLicenses) {
    vm.audios = audios.map((audio) => {
        if (audio.created_ago) {
            audio.created_ago = audio.created_ago.replace("&nbsp;", "\u00A0")
        }
        if (audio.modified_ago) {
            audio.modified_ago = audio.modified_ago.replace("&nbsp;", "\u00A0")
        }
        return audio
    })
    vm.audioLicenses = audioLicenses;
}
```

This way we don't have to worry about the security implications of using `trustAsHtml()` for any translations.

cblanken replied:
@jiru ^ would this be an acceptable solution?

jiru replied:
> This way we don't have to worry about the security implications of using trustAsHtml() for any translations.

Regardless of this PR, we already have to worry about the security implications of trusting translations as HTML in pretty much all the frontend code that is not AngularJS-ified. We do this by manually checking translated strings when committing `.po` files so it’s not a such big worry.

> would this be an acceptable solution?

Sorry but I don’t think this is a good way to approach the problem. I’d like to suggest 2 alternatives:
* leave the translatable string with `&nbsp;` and use `trustAsHtml()` (I can commit that specific change if you want to avoid responsibility for any security implication)
* change `&nbsp;` to `\u{00A0}` and I may be able to do some regexp magic on Transifex to keep all translations unaffected

cblanken replied:
Ah okay. I understand. If it's an implicit assumption throughout the app, you're right it doesn't really make a difference for this PR. As long as there is awareness of it. I'll go ahead and revert the `DateHelper` to keep using `&nbsp;` and use `trustAsHtml()` as recommended.

cblanken replied:
Okay, I've reverted to using the `&nbsp;`. I elected to use the `ng-bind-html` Angular attribute to allow the HTML rendering since it seemed more in line with the existing code than including the `$sce.trustAsHtml()`.

cblanken commented:
If there's anything else I need to do to finish up this PR, please let me know. I'm happy to fix any remaining issues.
